### DISCUSSION TITLE
Boundary ghost UO of two-phase flow

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#26502] [Q&A Modules: Thermal Hydraulics] [answered]
Hi all,

I noticed that the THM has` ADBoundaryFlux3EqnGhostVelocityTemperature` to calculate ghost variables of single-phase flow. In this file, if it is inlet flow, we use the input temperature. Otherwise, we use the interior temperature value. I wonder if it is two-phase flow (six-equation model), what does it should be? Calculating liquid ghost variables (p, ul, Tl) according to liquid velocity and calculating vapor ghost variables (alpha, ug, Tg) according to vapor velocity? Here is my code:

```
  params.addRequiredParam<Real>("alpha", "Specified void fraction");
  params.addRequiredParam<Real>("velocity_liq", "Specified liquid velocity");
  params.addRequiredParam<Real>("velocity_vap", "Specified vapour velocity");
  params.addRequiredParam<Real>("temperature_liq", "Specified liquid temperature");
  params.addRequiredParam<Real>("temperature_vap", "Specified vapour temperature");

    _alpha_ghost(getParam<Real>("alpha")),
    _velocity_liq_ghost(getParam<Real>("velocity_liq")),
    _velocity_vap_ghost(getParam<Real>("velocity_vap")),
    _temperature_liq_ghost(getParam<Real>("temperature_liq")),
    _temperature_vap_ghost(getParam<Real>("temperature_vap")),


  if (!_reversible || THM::isInlet(_velocity_liq_ghost, _normal))  
  {
    // Pressure is the only quantity coming from the interior
    W_ghost[THM::PRIM_VAR_PRESSURE] = p_interior;
    W_ghost[THM::PRIM_VAR_VEL_LIQ] = _velocity_liq_ghost;
    W_ghost[THM::PRIM_VAR_TEMP_LIQ] = _temperature_liq_ghost;
  }
  else  
  {
    // Pressure is the only quantity coming from the interior
    W_ghost[THM::PRIM_VAR_PRESSURE] = p_interior;
    W_ghost[THM::PRIM_VAR_VEL_LIQ] = _velocity_liq_ghost;
    W_ghost[THM::PRIM_VAR_TEMP_LIQ] = temp_liq_interior;
  }

  // vapour phase
  if (!_reversible || THM::isInlet(_velocity_vap_ghost, _normal))  
  {
    W_ghost[THM::PRIM_VAR_VOID_FRACTION] = _alpha_ghost;
    W_ghost[THM::PRIM_VAR_VEL_VAP] = _velocity_vap_ghost;
    W_ghost[THM::PRIM_VAR_TEMP_VAP] = _temperature_vap_ghost;
  }
  else 
  {
    W_ghost[THM::PRIM_VAR_VOID_FRACTION] = _alpha_ghost;
    W_ghost[THM::PRIM_VAR_VEL_VAP] = _velocity_vap_ghost;
    W_ghost[THM::PRIM_VAR_TEMP_VAP] = temp_vap_interior;
  }
```

Is there anything wrong? I'm not sure.

**Comment by SomnusYu** [comment]
@joshuahansel Could you please help me?

**Comment by joshuahansel** [comment, marked as answer]
In RELAP-7 (7-equation model), for checking reversibility, we check the sign of `arhouA_liquid + arhouA_vapor`, not individually by phase as you've done. However, what you're doing might be ok too. I don't really know. The situation is a bit more complicated than both of these approximations because there are a lot of waves involved. The objective should just be to supply what you know about the ghost state for your situation and have the Riemann solver figure out the appropriate information to use based on its knowledge of the characteristics.

**Reply by SomnusYu** [reply]
Thanks a lot! By the way, do you have any reference to deal with this situation :

> check the sign of arhouA_liquid + arhouA_vapor
